src/msg/producer/writer/message_writer.go

```diff
@@ -747,3 +755,76 @@ func (a *acks) size() int {
 	a.Unlock()
 	return l
 }
+
+type metricIdx byte
+
+const (
+	_messageClosed metricIdx = iota
+	_messageDroppedBufferFull
+	_messageDroppedTTLExpire
+	_messageRetry
+	_processedAck
+	_processedClosed
+	_processedDrop
+	_processedNotReady
+	_processedTTL
+	_processedWrite
+	_lastMetricIdx
+)
+
+type scanBatchMetrics [_lastMetricIdx]int32
+
+func (m *scanBatchMetrics) record(metrics *messageWriterMetrics) {
+	m.recordNonzeroCounter(_messageClosed, metrics.messageClosed)
+	m.recordNonzeroCounter(_messageDroppedBufferFull, metrics.messageDroppedBufferFull)
+	m.recordNonzeroCounter(_messageDroppedTTLExpire, metrics.messageDroppedTTLExpire)
+	m.recordNonzeroCounter(_messageRetry, metrics.messageRetry)
+	m.recordNonzeroCounter(_processedAck, metrics.processedAck)
+	m.recordNonzeroCounter(_processedClosed, metrics.processedClosed)
+	m.recordNonzeroCounter(_processedDrop, metrics.processedDrop)
+	m.recordNonzeroCounter(_processedNotReady, metrics.processedNotReady)
+	m.recordNonzeroCounter(_processedTTL, metrics.processedTTL)
+	m.recordNonzeroCounter(_processedWrite, metrics.processedWrite)
+}
+
+func (m *scanBatchMetrics) recordNonzeroCounter(idx metricIdx, c tally.Counter) {
+	if m[idx] > 0 {
+		c.Inc(int64(m[idx]))
+	}
+}
+
+func nextRetryNanosFn(retryOpts retry.Options) func(int) int64 {
+	var (
+		jitter              = retryOpts.Jitter()
+		backoffFactor       = retryOpts.BackoffFactor()
+		initialBackoff      = retryOpts.InitialBackoff()
+		maxBackoff          = retryOpts.MaxBackoff()
+		initialBackoffFloat = float64(initialBackoff)
+	)
+
+	// inlined and specialized retry function that does not have any state that needs to be kept
+	// between tries
+	return func(writeTimes int) int64 {
+		backoff := initialBackoff.Nanoseconds()
+		if writeTimes >= 1 {
+			backoffFloat64 := initialBackoffFloat * math.Pow(backoffFactor, float64(writeTimes-1))
+			backoff = int64(backoffFloat64)
+		}
+		// Validate the value of backoff to make sure Int63n() does not panic.
+		if jitter && backoff >= 2 {
+			half := backoff / 2
+			backoff = half
+			if backoff < math.MaxUint32 {
```

Comment:
Nit: maybe worth a comment with "// Prevent overflow. 
If backoff < math.MaxUint32, we know that backoff + jitter < math.MaxUint32 since we just divided backoff by 2"

Reply:
Or tbh not; I am kind of an overexplainer in comments. Take or leave it haha.